distribution/pull.go

```diff
@@ -165,11 +168,11 @@ func Pull(ref reference.Named, imagePullConfig *ImagePullConfig) error {
 // status message indicates that a newer image was downloaded. Otherwise, it
 // indicates that the image is up to date. requestedTag is the tag the message
 // will refer to.
-func writeStatus(requestedTag string, out io.Writer, sf *streamformatter.StreamFormatter, layersDownloaded bool) {
+func writeStatus(requestedTag string, out chan<- xfer.Progress, layersDownloaded bool) {
 	if layersDownloaded {
-		out.Write(sf.FormatStatus("", "Status: Downloaded newer image for %s", requestedTag))
+		out <- xfer.Progress{Message: "Status: Downloaded newer image for " + requestedTag}
 	} else {
-		out.Write(sf.FormatStatus("", "Status: Image is up to date for %s", requestedTag))
+		out <- xfer.Progress{Message: "Status: Image is up to date for " + requestedTag}
```

Comment:
@aaronlehmann I'm not going to pretend to understand that code but I think the concept is correct. Basically, you have a broadcast loop, that looks like this:

```go
for _, dst := range dsts {
    select {
    case dst <- msg:
    default:
        // skip destination
    }
}
```

Listeners must be waiting at the time of send, or the channel must be buffered, to get an update. This can also be accomplished by have a reliable destination (from [notifications](https://github.com/docker/distribution/blob/master/notifications/sinks.go#L48)):

```go
func (b *Broadcaster) Write(events ...Event) error {
	select {
	case b.events <- events:
	case <-b.closed:
		return ErrSinkClosed
	}
	return nil
}
```

On the other side is a goroutine that ensures the events are pulled from b.events and dispatched to a waiting `Sink`. The target `Sink.Write` function must not block. To provide that guarantee, we back the other side with a queue. We could just as well provide that guarantee by dropping anything that can't be immediately sent.